src/http/router.ts

```diff
@@ -1218,46 +1218,48 @@ export const getRouter = (client: CLClient, options: ICLClientOptions): express.
                     const conversationReference = BB.TurnContext.getConversationReference(activity)
                     if (!conversationReference) {
                         validity = CLM.TranscriptValidationResultType.TEST_FAILED
+                        break
                     }
-                    else {
+
+                    // Get session Id
+                    const sessionId = await result.memory.BotState.GetSessionIdAsync()
+                    if (!sessionId) {
+                        validity = CLM.TranscriptValidationResultType.TEST_FAILED
+                        break
+                    }
+
+                    if (conversationReference && sessionId) {
                         // Include apiResults when taking action so result will be the same when testing
                         await clRunner.TakeActionAsync(conversationReference, result, null, turnValidation.apiResults[0])
-                    }
 
-                    // Need to manually run through non-wait actions to retrieve actions to validate against
-                    let bestAction = result.scoredAction
-                    if (turnValidation.actionHashes.length > 1) {
-                        const sessionId = await result.memory.BotState.GetSessionIdAsync()
-                        if (!sessionId) {
-                            validity = CLM.TranscriptValidationResultType.TEST_FAILED
-                        }
-                        else {
-                            for (let i = 1; i < turnValidation.actionHashes.length; i = i + 1) {
-                                // If last action was terminal, can't do another action
-                                if (bestAction.isTerminal) {
-                                    validity = CLM.TranscriptValidationResultType.CHANGED
-                                }
-                                else {
-                                    bestAction = await clRunner.Score(appId, sessionId, result.memory, '', [], result.clEntities, false, true)
-                                    result.scoredAction = bestAction
-
-                                    // Did I select the expected action
-                                    if (turnValidation.actionHashes[i] && !Utils.actionHasHash(bestAction.actionId, turnValidation.actionHashes[i], appDefinition.actions)) {
-                                        validity = CLM.TranscriptValidationResultType.CHANGED
-                                    }
-                                    // Include apiResults when taking action so result will be the same when testing
-                                    await clRunner.TakeActionAsync(conversationReference, result, null, turnValidation.apiResults[i])
-                                }
+                        // Trigger next action if non-terminal
+                        let bestAction = result.scoredAction
+                        let curHashIndex = 0
+                        while (!bestAction.isTerminal) {
```

Comment:
Is it guaranteed that there will be a terminal action ?

Reply:
Not sure what you mean.

Reply:
Rephrasing - is there a possibility that this will loop infinitely ?

Reply:
Ah, gotcha.  The server has a max number of non-terminal actions that it allows.  I'll add a comment